### ISSUE TITLE
Include Export png from Visio

### ISSUE THREAD

**Issue by FrankPohl** [docToolchain/docToolchain#104]
Visio is often use to document parts of the architcture, so it makes sense to provide an export of png files for Visio documents. The Powershell scripts below does excatly this.
Would be great if this could be included in the toolchain.
```
# Convert all the pages in all visio files in the given folder to png or adoc files.
# A Visio windows might flash shortly.
# The converted png files are stored in the given folder
# The names of the png files are derived from the Viso file name and the page name.
# In addition to the png conversion all page comments are stored in files witj tje adoc extension.
# Parameter: Path where visio files can be found. If not given as command line parameter the script prompts for it
# Example: powershell VisoPageToPngConverter.ps1 -Path c:\convertertest\

Param
(
    [Parameter(Mandatory=$true,ValueFromPipeline=$true,Position=0)]
    [Alias('p')][String]$SourcePath
)

If (!(Test-Path -Path $SourcePath))
{
    Write-Warning "The path ""$SourcePath"" does not exist or is not accessible, plese input the correct path."
    Exit
}

# Get all possible Visio files 
$VisioFiles = Get-ChildItem -Path $SourcePath -Recurse -Include *.vsdx,*.vssx,*.vstx,*.vxdm,*.vssm,*.vstm,*.vsd,*.vdw,*.vss,*.vst

If(!($VisioFiles))
{
    Write-Warning "There are no Visio files in the path ""$SourcePath""."
    Exit
}

$VisioApp = New-Object -ComObject Visio.Application
$VisioApp.Visible = $false
    
# Extract the png and adoc from all the files in the folder
Foreach($File in $VisioFiles)
{
    $FilePath = $File.FullName
    $FileDirectory = $File.DirectoryName   #get the folder containing the Visio file. Will be used to store the png and adoc files
    $FileBaseName = $File.BaseName    # get the filename to be used as part of the name of the png and adoc files
    
    Try
   {
        $Document = $VisioApp.Documents.Open($FilePath)
        $Pages = $VisioApp.ActiveDocument.Pages
        $PageName
        Foreach($Page in $Pages)
        {
            $Page.Export("$FileDirectory\$($FileBaseName)-$($Page.Name).png") 
            $AllPageComments = ""
            ForEach($PageComment in $Page.Comments)
            {
                $AllPageComments += $PageComment.Text
            }
            If ($AllPageComments)
            {
                $stream = [System.IO.StreamWriter] "$FileDirectory\$($FileBaseName)-$($Page.Name).adoc"
                $stream.WriteLine($AllPageComments)
                $stream.close()
            }                    
        }
        $Document.Close()
    }
    Catch
    {
        Write-Warning "One or more visio page(s) in file ""$FilePath"" have been lost in this converting."
    }
}
$VisioApp.Quit()
```

**Comment by rdmueller** [comment]
that's great! Can you please do ma a favor?
I would like to have your code with you as commiter in the history.
Could you please fork this project (button on the upper right corner) and add you code as file in the `/docToolchain/tree/master/scripts` folder. Afterwards, create a pull request (standard settings will be ok).
This can all be done through the github web interface - no need to install anything on your machine.

I will then do the rest and include it in the docToolchain build.

Thanx!

**Comment by rdmueller** [comment]
btw: your script works like a charm - I was surprised to see that visio supports comments on elements.
Now the question for me is, if it makes sense to extend the script in the same way as the sparx script works:
- a note begins with a {adoc:filename} tag. This let's you export different comments into different files.
- when there are several diagrams in one visio file (that's possible, isn't it?), then those adoc files are generated in addition per diagram.

would like to hear your thoughts on that!

**Comment by FrankPohl** [comment]
Hello Ralf,
I’ll have a look at both of your mails after my holidays (very long weekend in NRW).
Frank

Von: Ralf D. Müller [mailto:notifications@github.com]
Gesendet: Donnerstag, 26. Oktober 2017 21:46
An: docToolchain/docToolchain <docToolchain@noreply.github.com>
Cc: FrankPohl <frankpohl@live.com>; Author <author@noreply.github.com>
Betreff: Re: [docToolchain/docToolchain] Include Export png from Visio (#104)


btw: your script works like a charm - I was surprised to see that visio supports comments on elements.
Now the question for me is, if it makes sense to extend the script in the same way as the sparx script works:

  *   a note begins with a {adoc:filename} tag. This let's you export different comments into different files.
  *   when there are several diagrams in one visio file (that's possible, isn't it?), then those adoc files are generated in addition per diagram.

would like to hear your thoughts on that!

**Comment by FrankPohl** [comment]
Hello,
just made a fork and added a pull request to integrate the Visio->png converter into the doctoolchain.
Please check if everything is OK because this is my first Github pull request.
Frank

Von: Ralf D. Müller [mailto:notifications@github.com]
Gesendet: Donnerstag, 26. Oktober 2017 21:46
An: docToolchain/docToolchain <docToolchain@noreply.github.com>
Cc: FrankPohl <frankpohl@live.com>; Author <author@noreply.github.com>
Betreff: Re: [docToolchain/docToolchain] Include Export png from Visio (#104)


btw: your script works like a charm - I was surprised to see that visio supports comments on elements.
Now the question for me is, if it makes sense to extend the script in the same way as the sparx script works:

  *   a note begins with a {adoc:filename} tag. This let's you export different comments into different files.
  *   when there are several diagrams in one visio file (that's possible, isn't it?), then those adoc files are generated in addition per diagram.

would like to hear your thoughts on that!

**Comment by rdmueller** [comment]
Great! Since you forked the repository from another fork, the pull request was create for this other repository.
I've initiated another PR from your fork to the master docToolchain repository.

You are now an official contributor!

https://github.com/docToolchain/docToolchain/graphs/contributors

Thanx for helping with this script. I will try to create the gradle part this week.

**Comment by rdmueller** [comment]
I'm nearly done with the integration but now I run into a problem with the powershell script.

when I execute it, it can't access the viso com object:

```
Ausnahme beim Festlegen von "Visible": "Das COM-Objekt des Typs "Microsoft.Office.Interop.Visio.ApplicationClass" kann nicht in den Schnittstellentyp
"Microsoft.Office.Interop.Visio.IVApplication" umgewandelt werden. Dieser Vorgang konnte nicht durchgeführt werden, da der QueryInterface-Aufruf an die
COM-Komponente für die Schnittstelle mit der IID "{000D0700-0000-0000-C000-000000000046}" aufgrund des folgenden Fehlers nicht durchgeführt werden konnte:
Bibliothek nicht registriert. (Ausnahme von HRESULT: 0x8002801D (TYPE_E_LIBNOTREGISTERED))."
In ...\scripts\VisioPageToPngConverter.ps1:48 Zeichen:1
+ $VisioApp.Visible = $false
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], SetValueInvocationException
    + FullyQualifiedErrorId : ExceptionWhenSetting
```

any idea? 
Google says "reinstall office" - but I am looking for a different solution.

source is available here: https://github.com/rdmueller/forked_docToolchain/tree/visio

**Comment by rdmueller** [comment]
Just managed to "repair" my visio installation and it works again... will explain this in the FAQ.

**Comment by rdmueller** [comment]
will be fixed by PR #114
thanx for your patience!

**Comment by rdmueller** [comment]
oh. this had already been fixed but not closed.
Thanx again for your PR!
